### PR TITLE
PPS detector mapping records separation and serialization

### PR DESCRIPTION
**Pull request by brzozia**
#### PR description:

This PR is a first step in a way to resolve an issue from https://github.com/cms-sw/cmssw/pull/40765 - Transfer of all PPS detectors' mappings from XML to CondDB.

In this PR:
* Introduction of new record: [TotemAnalysisMaskRcd.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-8e15d43975e75857b41d01b6bfd950e6e7a2a87b203ea58acce2cad3100e8490) - separated from TotemReadoutRcd;
* Serialization of classes which objects are intended to be placed in the CondDB (and inner classes of those) - [TotemAnalysisMask.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-0266d70e3769fc2f3d773fb4a84ae048b23986dcff928e5a65d104dfd87770b9), [TotemDAQMapping.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-c70098b419b2deac1b51936061ac94c788496fa4724139ff74c276006e161b56), [TotemFramePosition.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-a64176b2f94eec46393738ac968202e6055cc1795695836c6f5f28c62dd63e7b), [TotemSymbId.h](https://github.com/CTPPS/cmssw/pull/79/files#diff-d3e99bdc9a0a36bab8559221eb41b4797245dd34799c91c6aced88e3ce6853d4));
* Writer plugin ([WriteCTPPSTotemDAQMappingMask.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-140c17cd5e469a3e3692f7f3e397d10512b8b9e95d1038ac986f3737902f8475)) and config ([write-ctpps-totem_daqmap_cfg.py](https://github.com/CTPPS/cmssw/pull/79/files#diff-e836bfb51a0ad8796ba77e01102e7487caaedfa09c9d56a28f7b1c59d4b8b5d0)) for reading CTPPS XML files with mappings and placing objects made from those in SQLite;
* Updates of printer plugin ([PrintTotemDAQMapping.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-2f22f8e72a591e6be95ba3f5388a9a837648b9153fb5ba8e7bfec2b1cc2211ff)) which prints records (TotemReadoutRcd, TotemAnalysisMaskRcd) content;

#### PR validation:

Validation has been made with use of new plugin ([CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc](https://github.com/CTPPS/cmssw/pull/79/files#diff-276299009e7ad4e9336945c66d7cfcb5b6ec85d7835707bcd3096bcfb382c7dc)) which reads data from SQLite/XML file and drops them to text file.
All data from XML was dropped to SQLite in separate records and with different IOVs.
```
python3 CondTools/CTPPS/test/script-ctpps-write-many-XML-to-SQLite.py
```
Then (using new plugin) objects were read from SQLite and dropped to text file.
```
python3 CalibPPS/ESProducers/test/script_test_many_writeTotemDAQMapping.py
```
Also data from XML files was read and dropped to other text file (change flag inside script and then run).
```
python3 CalibPPS/ESProducers/test/script_test_many_writeTotemDAQMapping.py
```
Then files were compared and their content was the same, as expected eg.:
``` 
diff all_diamonds_db.txt all_diamonds_xml.txt
```

---

@grzanka for your attention

### CODE REVIEW

File: CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc [hunk 1]
```diff
@@ -30,33 +30,34 @@ class PrintTotemDAQMapping : public edm::one::EDAnalyzer<> {
...
+      maskToken(esConsumes(edm::ESInputTag("", subSystemName))) {}
 
 //----------------------------------------------------------------------------------------------------
 
 void PrintTotemDAQMapping::analyze(const edm::Event &, edm::EventSetup const &es) {
   // get mapping
-  auto const &mapping = es.getData(mappingToken_);
+  if (auto mappingHandle = es.getHandle(mappingToken)) {
+    auto const &mapping = *mappingHandle;
+    std::cout << mapping;
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc:49]: Please use `LogPrint("PrintTotemDAQMapping") instead

File: CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc [hunk 2]
```diff
@@ -30,33 +30,34 @@ class PrintTotemDAQMapping : public edm::one::EDAnalyzer<> {
...
-  for (const auto &p : mapping.VFATMapping)
-    edm::LogInfo("PrintTotemDAQMapping mapping") << "    " << p.first << " -> " << p.second;
-
-  // print mapping
-  for (const auto &p : analysisMask.analysisMask)
-    edm::LogInfo("PrintTotemDAQMapping mask") << "    " << p.first << ": fullMask=" << p.second.fullMask
-                                              << ", number of masked channels " << p.second.maskedChannels.size();
+  if (auto analysisMaskHandle = es.getHandle(maskToken)) {
+    auto const &analysisMask = *analysisMaskHandle;
+    std::cout << analysisMask;
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc:57]: same

File: CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+public:
+  WriteTotemDAQMapping(const edm::ParameterSet &ps);
+  ~WriteTotemDAQMapping() override {}
+
+private:
+  /// label of the CTPPS sub-system
+  std::string subSystemName;
+  std::ofstream outStream;
+  edm::ESGetToken<TotemDAQMapping, TotemReadoutRcd> mappingToken;
+  edm::ESGetToken<TotemAnalysisMask, TotemAnalysisMaskRcd> maskToken;
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc:33]: I think these could be `const`

File: CondFormats/DataRecord/interface/TotemAnalysisMaskRcd.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#ifndef TotemAnalysisMaskRcd_TotemAnalysisMaskRcd_h
```
Comment (tvami) [inline CondFormats/DataRecord/interface/TotemAnalysisMaskRcd.h:1]:
```suggestion
#ifndef CondFormats_DataRecord_TotemAnalysisMaskRcd_h
```

**Review by tvami — COMMENTED** [review]
The tests you mention in the PR description should become unit test, `cout` should be avoided, and I think the tokens can be `const`-s. Thanks!

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by grzanka** [comment]
@tvami I was looking at the failed tests, for example this one:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-3c9132/34177/runTheMatrixINPUT-results/14034.0_TTbar_14TeV+2023FS/step2_TTbar_14TeV+2023FS.log

```
/tmp/cmsbuild/jenkins/workspace/ib-run-pr-relvals/cms-bot/pr_testing/retry-command.sh: line 16: 28455 Killed                  $@
```

It doesn't seem to be related to the changes introduced in this PR, can you take a look ?

**Comment by tvami** [comment]
@cmsbuild, please test

File: CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc [hunk 5]
```diff
@@ -30,33 +30,34 @@ class PrintTotemDAQMapping : public edm::one::EDAnalyzer<> {
 private:
   /// label of the CTPPS sub-system
   std::string subSystemName;
-  edm::ESGetToken<TotemDAQMapping, TotemReadoutRcd> mappingToken_;
-  edm::ESGetToken<TotemAnalysisMask, TotemReadoutRcd> maskToken_;
+  edm::ESGetToken<TotemDAQMapping, TotemReadoutRcd> mappingToken;
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc:33]: I see you change from `mappingToken` to `mappingToken_` this is the opposite of what CMSSW naming rules suggest. See 2.9 in https://cms-sw.github.io/cms_coding_rules.html (and then technically the `subSystemName` should have been `subSystemName_`

Comment (tvami) [reply CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc:33]: Besides the cosmetics, these could be `const`, no?

File: CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc [hunk 6]
```diff
@@ -30,33 +30,34 @@ class PrintTotemDAQMapping : public edm::one::EDAnalyzer<> {
...
 //----------------------------------------------------------------------------------------------------
 
 void PrintTotemDAQMapping::analyze(const edm::Event &, edm::EventSetup const &es) {
   // get mapping
-  auto const &mapping = es.getData(mappingToken_);
+  if (auto mappingHandle = es.getHandle(mappingToken)) {
+    auto const &mapping = *mappingHandle;
+    edm::LogPrint("PrintTotemDAQMapping") << mapping;
+  } else {
+    edm::LogError("PrintTotemDAQMapping mapping") << "No mapping found";
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/PrintTotemDAQMapping.cc:51]: the body of the `edm::LogError` should have the name of the module, i.e. `PrintTotemDAQMapping`

File: CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc [hunk 7]
```diff
@@ -0,0 +1,67 @@
+/****************************************************************************
+ *
+ ****************************************************************************/
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc:3]: delete?

File: CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+public:
+  WriteTotemDAQMapping(const edm::ParameterSet &ps);
+  ~WriteTotemDAQMapping() override {}
+
+private:
+  /// label of the CTPPS sub-system
+  std::string subSystemName;
+  std::ofstream outStream;
+  const edm::ESGetToken<TotemDAQMapping, TotemReadoutRcd> mappingToken;
+  const edm::ESGetToken<TotemAnalysisMask, TotemAnalysisMaskRcd> maskToken;
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc:33]: thanks for having this `const`, the trailing `_` in the end would be appreciated too

File: CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+//----------------------------------------------------------------------------------------------------
+
+void WriteTotemDAQMapping::analyze(const edm::Event &, edm::EventSetup const &es) {
+  // get mapping
+  if (auto mappingHandle = es.getHandle(mappingToken)) {
+    auto const &mapping = *mappingHandle;
+    outStream << mapping;
+  } else {
+    edm::LogError("WriteTotemDAQMapping mapping") << "No mapping found";
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/WriteTotemDAQMapping.cc:51]: same comment for LogError as above

**Comment by brzozia** [comment]
Thank you for the review. 
I have implemented the suggested changes and enhanced the printing feature required for proper unit testing. 
I still need to work on the printouts for the T2 detector, but I hope to complete this along with the unit test soon. 
Sorry for the delay

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by grzanka** [comment]
> > I still need to work on the printouts for the T2 detector, but I hope to complete this along with the unit test soon.
> 
> Thanks! It's good at least that everything else worked for the `limited` test above

@tvami which tests exactly are being executed by cmsbot, you've mentioned here "limited". Is there as well a "full" test which can be triggered on Jenkins ?

**Comment by grzanka** [comment]
I've also seen that PR which enables retrieving conditions from HDF5 (https://github.com/cms-sw/cmssw/pull/42431/files) file was merged recently and that introduced a conflict here.
What is exactly `REGISTER_PLUGIN_NO_SERIAL` ? @tvami It it relevant for us for the scope of this PR ?

**Comment by brzozia** [comment]
I've finally added the unit test and made some (hopefully final) changes for printouts.
Its ready to be reviewed again :)

File: CondCore/CTPPSPlugins/src/plugin.cc [hunk 10]
```diff
@@ -48,5 +52,7 @@ REGISTER_PLUGIN(PPSDirectSimulationDataRcd, PPSDirectSimulationData);
 REGISTER_PLUGIN(PPSPixelTopologyRcd, PPSPixelTopology);
 REGISTER_PLUGIN(PPSAlignmentConfigRcd, PPSAlignmentConfig);
 REGISTER_PLUGIN(PPSAlignmentConfigurationRcd, PPSAlignmentConfiguration);
+REGISTER_PLUGIN(TotemAnalysisMaskRcd, TotemAnalysisMask);
+REGISTER_PLUGIN(TotemReadoutRcd, TotemDAQMapping);
```
Comment (grzanka) [inline CondCore/CTPPSPlugins/src/plugin.cc:56]: @tvami should we use `REGISTER_PLUGIN_NO_SERIAL` or `REGISTER_PLUGIN` ?

I am asking as I've seen many changes like that in recent HDF5 PR:
image

Comment (tvami) [reply CondCore/CTPPSPlugins/src/plugin.cc:56]: Let me tag @Dr15Jones about this

Comment (francescobrivio) [reply CondCore/CTPPSPlugins/src/plugin.cc:56]: If the same class is mapped to multiple records:
  - One registering should use `REGISTER_PLUGIN`
  - All the others should use `REGISTER_PLUGIN_NO_SERIAL`
 
In any case, IIUC the code won't compile if you have implemented it wrongly.
Chris, please correct me if I said something wrong.

**Comment by Dr15Jones** [comment]
> In any case, IIUC the code won't compile if you have implemented it wrongly. Chris, please correct me if I said something wrong.

that is correct, it won't compile.

**Comment by grzanka** [comment]
@tvami I think we are ready to accept more comments on this PR, would it be possible to run the tests again ?

**Comment by perrotta** [comment]
please test

File: CalibPPS/ESProducers/test/test_printTotemDAQMapping.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    timetype = cms.string('runnumber'),
+    firstValue = cms.uint64(340000),
+    lastValue = cms.uint64(340000),
+    interval = cms.uint64(1)
+)
+
+# load a mapping from XML
+process.load("CalibPPS.ESProducers.totemDAQMappingESSourceXML_cfi")
+process.totemDAQMappingESSourceXML.subSystem = "TimingDiamond"
+process.totemDAQMappingESSourceXML.sampicSubDetId = cms.uint32(6)
```
Comment (perrotta) [inline CalibPPS/ESProducers/test/test_printTotemDAQMapping.py:23]:
```suggestion
process.totemDAQMappingESSourceXML.sampicSubDetId = 6
```

**Review by perrotta — COMMENTED** [review]
Very minor comments, not even worth implementing unless there are other updates needed here

**Comment by grzanka** [comment]
> Very minor comments, not even worth implementing unless there are other updates needed here

From our side we do not foresee any updates, basically we are waiting for PR comments and approval

**Comment by perrotta** [comment]
As far as I can see this PR is good to go. The new unit test can run and succeed.
Unless there are other comments (e.g. @tvami from vacation...) I can sign for @cms-sw/alca-l2

**Comment by perrotta** [comment]
+alca,db

**Comment by perrotta** [comment]
+db

**Comment by perrotta** [comment]
+1
- [This](https://github.com/cms-sw/cmssw/pull/42471#pullrequestreview-1593204133) could be followed up in a forthcoming clean up PR

**Pull request merged**